### DISCUSSION TITLE
Running Gogs with rootless Podman

### DISCUSSION THREAD

**Discussion by eriksjolund** [gogs/gogs#6775] [General]
I wrote a Bash script for starting Gogs with rootless Podman:

```
#!/bin/bash
uid=1000
gid=1000
subuidSize=$(( $(podman info --format "{{ range .Host.IDMappings.UIDMap }}+{{.Size }}{{end }}" ) - 1 ))
subgidSize=$(( $(podman info --format "{{ range .Host.IDMappings.GIDMap }}+{{.Size }}{{end }}" ) - 1 ))
podman run --rm \
  -v ./gogs:/data:Z \
  --uidmap $uid:0:1 \
  --uidmap 0:1:$uid \
  --uidmap $(($uid+1)):$(($uid+1)):$(($subuidSize-$uid)) \
  --gidmap $gid:0:1 \
  --gidmap 0:1:$gid \
  --gidmap $(($gid+1)):$(($gid+1)):$(($subgidSize-$gid)) \
  --name=gogs -p 10022:22 -p 10880:3000 \
  docker.io/gogs/gogs
```
Let's run the script:
```
[tester@laptop ~]$ podman --version
podman version 3.4.4
[tester@laptop ~]$ mkdir gogs
[tester@laptop ~]$ bash run.sh 
usermod: no changes
Feb 18 20:01:26 syslogd started: BusyBox v1.33.1
2022/02/18 20:01:26 [ WARN] Custom config "/data/gogs/conf/app.ini" not found. Ignore this warning if you're running for the first time
2022/02/18 20:01:26 [TRACE] Log mode: Console (Trace)
2022/02/18 20:01:26 [ INFO] Gogs 0.13.0+dev
2022/02/18 20:01:26 [TRACE] Work directory: /app/gogs
2022/02/18 20:01:26 [TRACE] Custom path: /data/gogs
2022/02/18 20:01:26 [TRACE] Custom config: /data/gogs/conf/app.ini
2022/02/18 20:01:26 [TRACE] Log path: /app/gogs/log
2022/02/18 20:01:26 [TRACE] Build time: 2022-02-14 02:18:07 UTC
2022/02/18 20:01:26 [TRACE] Build commit: 8a1a40ce6a2fcad2ca877c1c98dcf492c5f9fbed
2022/02/18 20:01:26 [ INFO] Run mode: Development
2022/02/18 20:01:26 [ INFO] Listen on http://0.0.0.0:3000
Feb 18 20:01:26 sshd[45]: Server listening on :: port 22.
Feb 18 20:01:26 sshd[45]: Server listening on 0.0.0.0 port 22.
```

It seems to work, but I haven't tested so much. At least I see a web page 
on localhost:10880 and I see that it's possible to connect with an SSH client to localhost:10022.

```
[tester@laptop ~]$ ssh -p 10022 localhost
The authenticity of host '[localhost]:10022 ([::1]:10022)' can't be established.
ED25519 key fingerprint is SHA256:rRXoPCat1mA7cGWCr9TOUxGYzQafFUEr2yXKIH+Oe8w.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '[localhost]:10022' (ED25519) to the list of known hosts.
tester@laptop: Permission denied (publickey,keyboard-interactive).
[tester@laptop ~]$ 
```

The bind-mounted directory contains a few directories that are owned by the regular UID and GID

```
[tester@laptop ~]$ ls -l gogs
total 4
drwxr-xr-x. 3 tester tester   18 Feb 18 19:14 git
drwxr-xr-x. 5 tester tester   41 Feb 18 19:14 gogs
drwx------. 2 tester tester 4096 Feb 18 19:14 ssh
[tester@laptop ~]$ 
```

(Without __--uidmap__ and  __--gidmap__, the directories
would have been owned by a subordinate UID and GID)

Is there any recommended way of how to run Gogs with rootless Podman?

It's convenient to have the bind-mounted directories and files owned by the regular user, but 
maybe it's more secure to be running with a non-root user inside the container that is not mapped
to the regular user.

In the example Bash script above I used `--user 0:0` (the default value).
Is it it possible to start the Gogs container to run as a non-root user, e.g. by providing `--user 1000:1000`?
(a number greater than 0)

**Comment by unknwon** [comment]
Hey @eriksjolund, glad to see you've got it working so far and thanks for posting! Haven't had any to play with Podman yet so nice to see it is actually possible to run Gogs 😁 

>Is there any recommended way of how to run Gogs with rootless Podman?

We don't have any yet.

>Is it it possible to start the Gogs container to run as a non-root user, e.g. by providing --user 1000:1000?

I can't see why not, the Docker container is [using 1000:1000](https://github.com/gogs/gogs/blob/4ca87057f3da5d3b7befb8c2338288c7c1ec1a84/docker/start.sh#L47-L48).
